enigma-principal/app/src/epochs/verifier.rs

```
@@ -0,0 +1,208 @@
+use std::{
+    fs::{self, File},
+    io::prelude::*,
+    ops::Deref,
+    path::PathBuf,
+    sync::Mutex,
+};
+use std::clone::Clone;
+use std::sync::MutexGuard;
+use std::mem::replace;
+
+use rmp_serde::{Deserializer, Serializer};
+use serde::{Deserialize, Serialize};
+use web3::types::U256;
+
+use enigma_tools_m::keeper_types::EPOCH_CAP;
+
+use epochs::epoch_types::{ConfirmedEpochState, SignedEpoch};
+use esgx::general::{EPOCH_DIR, EPOCH_FILE};
+use common_u::custom_errors::{VerifierError, EpochError};
+
+#[derive(Debug)]
+pub struct EpochVerifier {
+    pub epoch_list: Mutex<Vec<SignedEpoch>>,
+    state_path: PathBuf,
+}
+
+impl EpochVerifier {
+    pub fn new(mut state_path: PathBuf) -> Result<Self, VerifierError> {
+        state_path.push(EPOCH_DIR);
+        if !state_path.exists() {
+            fs::create_dir_all(&state_path).or( Err(VerifierError::CreateErr))?;
+        }
+        state_path.push(EPOCH_FILE);
+        Self::create_from_path(state_path)
+    }
+    /// create an EpochVerifier object given the path from which we retrieve
+    /// the epoch states and we use the cap amount of epochs as well to handle backwards.
+    fn create_from_path(state_path: PathBuf) -> Result<EpochVerifier, VerifierError> {
+        let signed_epochs = match File::open(&state_path) {
+            Ok(mut f) => {
+                let mut buf = Vec::new();
+                f.read_to_end(&mut buf).or(Err(VerifierError::ReadErr))?;
+                let mut des = Deserializer::new(&buf[..]);
+                let mut data: Vec<SignedEpoch> = Deserialize::deserialize(&mut des).unwrap_or_default();
```

This is dangerous, as it silently ignores badly formatted files.